minique/models/priority_queue.py

```diff
@@ -0,0 +1,138 @@
+        job_index = insert_status
+    end
+else
+    -- The queue is empty.
+    insert_status = redis.call("RPUSH", queue_key, job_id)
+end
+
+if insert_status == -1 then
+    -- Our place in the queue is somewhere in the middle, search for it
+    for i, queued_job_id in ipairs(redis.call("LRANGE", queue_key, 0, -1)) do
+        local queued_job_priority = tonumber(redis.call("HGET", queue_prio_lookup, queued_job_id)) or 0
+        if queued_job_priority < job_priority then
+            -- Found the first job with a lower priority, insert before it
+            insert_status = redis.call("LINSERT", queue_key, "BEFORE", queued_job_id, job_id)
+            job_index = i
+            break
+        end
+    end
+end
+
+if insert_status == -1 then
+    -- No higher priority jobs found, insert to the end
+    redis.call("RPUSH", queue_key, job_id)
+end
+
+-- Introduce the newly added job to the priority lookup hash
+redis.call("HSET", queue_prio_lookup, job_id, job_priority)
+
+return job_index
+"""
+
+PRIO_HASH_CLEANER_SCRIPT = """
+local queue_key = ARGV[1]
+local queue_prio_lookup = ARGV[2]
+
+local queue_len = tonumber(redis.call("LLEN", queue_key))
+local hash_len = tonumber(redis.call("HLEN", queue_prio_lookup))
+
+-- Tolerate some excess
+if queue_len + 100 >= hash_len then
+    return -1
+end
+
+local new_hash = {}
+local count = 0
+
+for i, queued_job_id in ipairs(redis.call("LRANGE", queue_key, 0, -1)) do
+    local priority = redis.call("HGET", queue_prio_lookup, queued_job_id)
+    new_hash[queued_job_id] = priority
+    count = i
+end
+redis.call("DEL", queue_prio_lookup)
+
+for queued_job_id, priority in pairs(new_hash) do
+    redis.call("HSET", queue_prio_lookup, queued_job_id, priority)
+end
+
+return hash_len - count
+"""
+
+
+class PriorityQueue(Queue):
+    """Alternate Queue implementation that uses the optional `Job.priority` attribute
+    to sort the queue when adding jobs.
+
+    When using the priority queue, ensure the `PriorityQueue.finish_job()` method is called
+    for each job after it leaves the queue to trim the job priority lookup hash.
+    This can be performed by either the consumer or the job manager, or both.
+
+    Alternately, or in addition, periodically call `PriorityQueue.periodic_clean()` to
+    remove stale keys from the priority lookup hash.
+    """
+
+    def __init__(self, redis: "Redis[bytes]", name: str):
+        super().__init__(redis, name)
+        self.add_job_script = redis.register_script(ADD_JOB_SCRIPT)
+        self.hash_clean_script = redis.register_script(PRIO_HASH_CLEANER_SCRIPT)
+
+    @property
+    def prio_key(self) -> str:
+        return f"{self.redis_key}prio"
+
+    def add_job(self, job: "Job") -> int:
+        script_response = self.add_job_script(
+            keys=[self.redis_key, self.prio_key, job.redis_key],
+            args=[self.redis_key, self.prio_key, job.redis_key, job.id],
+        )
+        return cast(int, script_response) - 1
```

Comment:
Shouldn't just
```suggestion
        return int(script_response) - 1
```
do?